File: 2494-SumOfPrefixScoresOfStrings/2494-SumOfPrefixScoresOfStrings.java
// Last updated: 10/20/2025, 11:35:20 AM
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    int count = 0;
}

class Trie {
    TrieNode root = new TrieNode();
    
    void insert(String s) {
        TrieNode node = root;

        for (char c : s.toCharArray()) {
            if (node.children[c - 'a'] == null) {
                node.children[c - 'a'] = new TrieNode();
            }
            node = node.children[c - 'a'];
            node.count++;
        }
    }

    int findScore(String s) {
        TrieNode node = root;
        int sum = 0;
        for (char c : s.toCharArray()) {
            node = node.children[c - 'a'];
            sum += node.count;
        }
        return sum;
    }
}

class Solution {
    public int[] sumPrefixScores(String[] words) {
        Trie root = new Trie();

        for (String str : words) {
            root.insert(str);
        }

        int[] res = new int[words.length];
        for (int i = 0; i < res.length; i++) {
            res[i] = root.findScore(words[i]);
        }
        return res;
    }
}

File: 2494-SumOfPrefixScoresOfStrings/2494-SumOfPrefixScoresOfStrings.py
# Last updated: 10/20/2025, 11:35:15 AM
# Constant for the number of letters in the alphabet
ALPHABET_SIZE = 26

class TrieNode:
    def __init__(self):
        # Each TrieNode has children representing 26 letters and a count for tracking prefixes
        self.children = [None] * ALPHABET_SIZE
        self.count = 0

class Trie:
    def __init__(self):
        # Initialize the root of the Trie
        self.root = TrieNode()
    
    def insert(self, word):
        current_node = self.root
        for char in word:
            index = ord(char) - ord('a')
            # Create a new TrieNode if the path doesn't exist
            if not current_node.children[index]:
                current_node.children[index] = TrieNode()
            current_node = current_node.children[index]
            current_node.count += 1  # Increment the prefix count
    
    def find_score(self, word):
        current_node = self.root
        total_score = 0
        for char in word:
            index = ord(char) - ord('a')
            current_node = current_node.children[index]
            total_score += current_node.count
        return total_score

class Solution:
    def sumPrefixScores(self, words):
        """
        :type words: List[str]
        :rtype: List[int]
        """
        # Build the Trie from the list of words
        trie = Trie()
        for word in words:
            trie.insert(word)

        # Use a list comprehension to calculate scores for each word
        return [trie.find_score(word) for word in words]
